### ISSUE TITLE
free@home Binary Sensor units are not recongized in HA

### ISSUE THREAD

**Issue by SuperMaximus1984** [jheling/freeathome#37]
Hi jheling!
Thanks for your efforts on creation of F@H module for HA.
I have FW 2.4.0 and almost all Actuators/Sensors discovered (modified pfreeathome.py with IDs of my devices that have been discovered by Openhab), however I am missing several important modules.
Here are Binary sensor modules with 4 input channels each:
ABB265232951 (KTQ)
ABB240232951 (PEH)
and partial problems with 8/8 Actuator / Binary input unit (ABB26B703051 (ZWX)): Here Actuator outputs are discovered, though Input channels are not discovered.
8/8 Actuator/Binary Input unit code is "B008", I tried inputting it as Sensor/Acutator in configuration file, also attempted adding it as Binary Sensor unit, but no result. I have some devices discovered in HA that seem to be correct, but they are always in OFF state.
Besides, all important Binary sensors are connected to 
ABB265232951 (KTQ)
ABB240232951 (PEH)
that are not discovered at all.

I was unable to run "get-master-message.py", was getting error

D:\>python get-master-message.py
Traceback (most recent call last):
  File "get-master-message.py", line 26, in <module>
    from fah.messagereader import MessageReader
ModuleNotFoundError: No module named 'fah'

Please help to sort things out and get Binary sensor modules working.
Thank you a lot!

**Comment by SuperMaximus1984** [comment]
UPD:
8/8 Actuator / Binary unit is discovered, but because it was input as Sensor/Actuator in .PY config file and I found out INPUTs were also discovered, but as Switches. Surely they don't work as Binary Switches and it's frustrating. How can I divide one device into 2: Acutator & Binary Sensor?
![image](https://user-images.githubusercontent.com/62514634/77256092-b2344e80-6c74-11ea-9ca3-6010af69646c.png)

**Comment by jheling** [comment]
If you want to run the get-master-message.py you have to place the full freeathome  directory somewhere on your computer and then run get-master-message.py within that directory. I use idle (python 3.7) on windows 10 for running get-master-message.py 

I know the binary sensors are not working. I have allready made some modifications. I'll try to upload them this week. 

Did you add new device_id's to your pfreeathome.py?  Can you share them? Other users could also benefit from them.

**Comment by SuperMaximus1984** [comment]
Sorry, but I did not manage to get "get-master-message.py" working even if I execute it from freeathome directory with original file structure. Firstly there was absent lib, I installed it, then one more error again.

> D:\freeathome>python get-master-message.py

Traceback (most recent call last):
  File "get-master-message.py", line 27, in <module>
    from fah.saslhandler import SaslHandler
  File "D:\freeathome\fah\saslhandler.py", line 7, in <module>
    from .crypto import Crypto, loginSaslPayload, buildSaslResponse
  File "D:\freeathome\fah\crypto.py", line 3, in <module>
    import libnacl
ModuleNotFoundError: No module named 'libnacl'

D:\freeathome>pip install libnacl
Collecting libnacl
  Downloading libnacl-1.7.1-py2.py3-none-any.whl (20 kB)
Installing collected packages: libnacl
Successfully installed libnacl-1.7.1

D:\freeathome>python get-master-message.py
Traceback (most recent call last):
  File "get-master-message.py", line 27, in <module>
    from fah.saslhandler import SaslHandler
  File "D:\freeathome\fah\saslhandler.py", line 7, in <module>
    from .crypto import Crypto, loginSaslPayload, buildSaslResponse
  File "D:\freeathome\fah\crypto.py", line 3, in <module>
    import libnacl
  File "C:\Users\Max\AppData\Local\Programs\Python\Python38\lib\site-packages\libnacl\__init__.py", line 85, in <module>
    nacl = _get_nacl()
  File "C:\Users\Max\AppData\Local\Programs\Python\Python38\lib\site-packages\libnacl\__init__.py", line 34, in _get_nacl
    raise OSError(msg)
OSError: Could not locate nacl lib, searched for libsodium

D:\freeathome>
> 

Yes, I added those device codes that were discovered by Openhab, I modified your code with my device codes and here it is:

>                      # Switch actuators
                if (device_id == 'B002' or device_id == '100E' or device_id == 'B008' or
                        device_id == '900D' or device_id == '9011' or device_id == '4000' or
                        device_id == '4001' or device_id == '100C' or device_id == '1010'):
                    self.add_light_device(neighbor, serialnumber, roomnames)

                # Dimming actuators
                # Hue Aktor (LED Strip), Sensor/dimaktor 1/1-voudig
                if (device_id == '101C' or device_id == '1021' or
                        device_id == '1014' or device_id == '901c' or
                        device_id == '9018' or device_id == '9019' or
                        device_id == '10C0' or device_id == '1017' or
                        device_id == '1019'):
                    self.add_dimmer_device(neighbor, serialnumber, roomnames)

                # Scene or Timer
                if device_id == '4800' or device_id == '4A00':
                    self.add_scene(neighbor, serialnumber, roomnames)

                # blind/cover device
                if device_id == 'B001' or device_id == '1013' or device_id == '1015' or \
                        device_id == '9014' or device_id == '9015':
                    self.add_cover_device(neighbor, serialnumber, roomnames)

                # Sensor units 1/2 way
                if device_id == '1002' or device_id == '1000' or device_id == '100C' or \
                    device_id == '1019' or device_id == '1017' :
                    self.add_sensor_unit(neighbor, serialnumber, roomnames, device_id)

                # binary sensor
                if device_id == 'B007':
                    self.add_binary_sensor(neighbor, serialnumber, roomnames)

                # movement detector
                if device_id == '100A' or device_id == '9008' or device_id == '900B':
                    self.add_movement_detector(neighbor, serialnumber, roomnames)

                # thermostat
                if device_id == '9005' or device_id == '2041':
                    self.add_thermostat(neighbor, serialnumber, roomnames)

For your convenience I add up codes below.
Binary sensor units were not recognized even by Openhab f@h module, so I just manually assigned "B007", though it's not discoverable. So still I hope I could run get-master-message.py and find out exact codes/IDs.

Binary input
B007
Blind   actuator sensor 1/1
9001
9014
Heating   actuator 6gang
B003
Movement   detector
9009
900B
Room   temperature Controller
9005
Room   temperature Controller Wireless
2041
Sensor   unit 1gang
9001
Sensor   unit 2gang
9003
Sensor/dimming   actuator
9018
Sensor/switch   actuator
B008
Switch   actuator sensor 1/1
900D
Switch   actuator sensor 2/2
9011
SysAP
1012
SysAP   communication interface
2012

**Comment by SuperMaximus1984** [comment]
After I installed libsodium, I got the following while running get-master-message.py

D:\freeathome>python get-master-message.py
INFO      2.4.0 10000 OuCskclpfkoPmW7fdpXf9tdBCUnoy4LjquEXy1M24pQ=
INFO      version: 2.4.0
From cffi callback <function _sock_state_cb at 0x000002779F2B1C10>:
Traceback (most recent call last):
  File "C:\Users\Max\AppData\Local\Programs\Python\Python38\lib\site-packages\pycares\__init__.py", line 91, in _sock_state_cb
    sock_state_cb(socket_fd, readable, writable)
  File "C:\Users\Max\AppData\Local\Programs\Python\Python38\lib\site-packages\aiodns\__init__.py", line 104, in _sock_state_cb
    self.loop.add_reader(fd, self._handle_event, fd, READ)
  File "C:\Users\Max\AppData\Local\Programs\Python\Python38\lib\asyncio\events.py", line 501, in add_reader
    raise NotImplementedError
NotImplementedError

**Comment by jheling** [comment]
There is also a libsodium.dll which should be installed in the Windows\SysWOW64 or Windows\System32. This dll does all the crypto computations.  The name of this dll should be libsodium.dll.

**Comment by SuperMaximus1984** [comment]
Yes, I copied this DLL to both directories and keep on getting this error in my previous message...

**Comment by jheling** [comment]
This error is new to me. Maybe install packaging with pip?

**Comment by SuperMaximus1984** [comment]
Yes, I did "pip install packaging" in the beginning.
You see it displays SysAP 2.4.0, but then fall back into some error when addresses to events.py.

Do I really need to run get-master-message.py in my case if you plan to implement Binary sensors in the nearest release?
Thank you!

**Comment by jheling** [comment]
No, it is not necessery. The error seems to be releated to python 3.8.  In version 3.8 asyncio uses a new event loop. https://stackoverflow.com/questions/58422817/jupyter-notebook-with-python-3-8-notimplementederror.
Reverting to 3.7 or patching a file in asyncio seems to be the solution.

**Comment by SuperMaximus1984** [comment]
Thanks! It worked!
As it was supposed my 4-channel Binary Input modules are having code "B007" and occasionally they were added to Objects list when I put "B007" in Switch/Actuator section, though they were not operational. Now B007 code is in right place (Binary Sensor), so I guess devices can be normally detected as soon as you release the update, right?

**Comment by jheling** [comment]
I have just updated the repository.  The B007 device is added  to the list of binary devices.

**Comment by SuperMaximus1984** [comment]
Thank you! 
What files should I replace in my current version?

**Comment by jheling** [comment]
binary_sensor.py and pfreeathome.py.

**Comment by SuperMaximus1984** [comment]
Thank you so much & big respect to you!
It DOES work!
Binary Sensor units are discoverable and operable.

Do you have an idea how I can make my B008 (which is a combined unit 8/8 Actuator/Binary sensor) discoverable not only as Actuator, but also as Binary module so that I could program HA actions based on inputs either?
Is it OK to add B008 to both sections of pfreeathome.py (Sensors and at the same time to Binary inputs)?

**Comment by SuperMaximus1984** [comment]
@jheling 
Dear Author, please clarify how to make B008 Sensor/ Schaltaktor 8/8fach, REG recognized by HA either like Actuator (now working like this) and Binary Input?
Unfortunately, when I input Binary Input 'B008' alongside with 'B008' as Sensor/Actuator, compenent stops working.
Now as Actuator inputs are recognized like Switches, and surely are not operable.
Thank you in advance!

![image](https://user-images.githubusercontent.com/62514634/78176981-0a820200-7466-11ea-8aac-a6b314662147.png)

**Comment by jheling** [comment]
In the current solution a device / channel combination cannot function as a binary input and  as a actuator. The software has evolved in a certain way and it is now not possible to combine both. I the future it could be possible, but it would take more changes in the software.

**Comment by SuperMaximus1984** [comment]
@jheling 
Would you please advise when will it be possible to combine binary input + actuator device and see all inputs/outputs in HA?
Pretty much needed feature...
Thanks a lot!

**Comment by Cavekeeper** [comment]
@jheling 
I have a problem due to 4x binary input sensor.
The binary sensor are recognized by HA, but not reacting.
The signal is always "high" although the input is configured as a normally open contact. A motion detector is connected to this. However, HA is not doing. The input always remains on "high".
In BJ FAH V2.5.5 the motion detector is working well and reacts to a simple action rule.
But now I want also to have this motion detector via binary input inside of HA.

![Bildschirmfoto 2020-10-23 um 15 09 01](https://user-images.githubusercontent.com/61421138/97009029-e2620980-1543-11eb-8566-5a5d787d0681.png)

Now I've also tried other normal switches. These are displayed in HA as "low" and then also react when the switch is pressed.

![Bildschirmfoto 2020-10-23 um 15 25 38](https://user-images.githubusercontent.com/61421138/97009151-0c1b3080-1544-11eb-88ee-94ab96872657.png)

Generally the binary switches seem to work. Only the 4-way binary input does not work in HA.
Where could be the problem?

*** RPI 3B+ | HassOS 4.15 | HA 0.116.4 | FAH V2.5.5 | last version of FAH component ***

**Comment by jheling** [comment]
Can you try to run get-master-message.py ?  The messages above tell you how to do that. With the mastermessage.xml I can see the configuration of the binary sensors.  Maybe something has changed in the 2.5.5 sysap and is the software looking at the wrong output datapoint.

**Comment by Cavekeeper** [comment]
Yes, here is the message:

`Thomass-MBP:freeathome tlintzhoeft$ python3 get-master-message.py
Traceback (most recent call last):
  File "/Users/tlintzhoeft/Downloads/freeathome/get-master-message.py", line 10, in <module>
    import slixmpp
ModuleNotFoundError: No module named 'slixmpp'
`

**Comment by jheling** [comment]
This get-master-message.py generates a mastermessage.xml and that is where I'm interested in.  This is a python error.  You have to install the slixmpp library and probably the libnacl library: 
`pip install slixmpp`
`pip install libnacl`

**Comment by Cavekeeper** [comment]
I also noticed the following. If I reconfigure the binary channel in BJ FAH, it is marked as OFF in HA. Then when the first trigger comes, the channel in HA goes ON and stays ON. It does not go OFF when the trigger is no longer present. That means in principle the trigger arrives in HA. It's ON. It just doesn't go back to OFF when the trigger drops.

**Comment by jheling** [comment]
Maybe there is something wrong in the message comming from FAH.  
Could you  turn the logging of HA to info?
`logger:
  default: info 
`
Then restart HA and  activate one of the movement detectors connnected to your binary sensor.
The initial value and actions should now appear in the HA logging.

**Comment by Cavekeeper** [comment]
I think I got a little closer to the problem.
If I configure the digital input in FAH as a switch, then the channel in HA goes ON and OFF with every trigger.
If I configure the digital input in FAH as a motion detector, then the channel in HA goes ON with the first trigger and remains there.
This is what it says in the LOG.
When configured as a switch, the channel changes from 0 to 1 and vice versa with each trigger.
When configured as a BWM, the channel remains at 1 with the first trigger.
I now believe that there must be a problem in the freeathome custom components after all.

**Comment by miezie** [comment]
Maybe wait for [this one](https://github.com/jheling/freeathome/pulls) from @Tho85 and then be happy :)

**Comment by jheling** [comment]
Can you share the logging of the testcases? Maybe there is something wrong with the messages coming from the sysap

**Comment by Cavekeeper** [comment]
Ok, this is the log by using the motion detector:

`2020-10-24 13:17:47 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x6e61b500> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:17:47 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 1
2020-10-24 13:17:49 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x690e71e0> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:17:49 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 1
2020-10-24 13:17:53 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x665f62f8> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:17:53 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 1
2020-10-24 13:17:54 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x690e7230> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:17:54 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 1
2020-10-24 13:17:58 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x66691780> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:17:58 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 1
2020-10-24 13:18:00 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x6daef8c0> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:18:00 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 1
`

And this is log after removing the motion detector:

`2020-10-24 13:21:11 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary info channel <Element 'channel' at 0x66ddbc08> device ABB2F2B11551/ch0002 in/output odp000D 
2020-10-24 13:21:11 INFO (MainThread) [custom_components.freeathome.pfreeathome] binary device ABB2F2B11551/ch0002 output odp000D is 0
`

**Comment by jheling** [comment]
The custom_domponent is following the messages from the sysap.  If the movement detector mode is active is only receiving on messages and is acting after that. 
I see that the movement sensor also has a parameter: Nopen and Nclose. Maybe you should change this parameter.
Or is the binary channel connected to a switch? Then you have the parameter "Switch-off delay" which can get in your way.

**Comment by Cavekeeper** [comment]
That's right, the motion detector and the digital input are configured as normally open contacts. On the BJ side, this button also works perfectly.
You are right, the signal also arrives in HA. It just doesn't drop out again when the motion detector no longer registers any movement.
This means that the binary input in HA must receive an automatic reset and go to 0.
The only problem is that there is no service in HA to change the status of a binary input.

**Comment by Tho85** [comment]
During my work in #73 I just tested a binary sensor 2-gang configured as a motion detector. I can confirm that the sensor detects the "on" signal, but then remains there. According to my debug logs, this is actually a limitation of free@home. The system never sends an "off" signal for movement detectors. BTW, I observe the same behaviour when I configure the input as "timed start/stop", i.e. staircase light sensor.

Another hint this is an issue with free@home is in the free@home UI itself. When you first configure a binary sensor as a movement detector, drop it onto the floor plan, connect it to a light and trigger a movement, then you should see the detector blinking shortly on the floor plan (this works for all binary sensors). But it does so only once, immediately after initial configuration. It then never lights up again when movement stops, or when movement is triggered a second time. Just the same as observed with Home Assistant.

TL;DR: It's most likely a free@home issue.

**Comment by miezie** [comment]
Is it an issue? You can just observe the lux value of the detector... it toggles based on movement and that value. You can still check if the specific light (or whatever) is on.

Actually depending on on/off state of a binary switch might also be weird because it simulates a switch. But it could also be just one button with a state toggle every time when you press it.

**Comment by Tho85** [comment]
@miezie This works for dedicated in-wall movement detectors, because they have a brightness sensor included. The issue is with generic binary sensors like [this one](https://search.abb.com/library/Download.aspx?DocumentID=2CDC550002D0101&LanguageCode=de&DocumentPartId=&Action=Launch), where you essentially connect an external movement detector to a generic input and configure it as a movement detector in F@H. They do not offer a brightness sensor.

> Actually depending on on/off state of a binary switch might also be weird because it simulates a switch. But it could also be just one button with a state toggle every time when you press it.

Yes, relying on the internal state of the binary sensor is flawed, even if you configure it as a toggle. The state will eventually mismatch if you change the state of the attached light in HA. I think the best solution if you want to trigger something in HA from a F@H binary sensor is to expose the HA entity via emulated_hue, e.g. as a light bulb. This way the state tracking in F@H is always correct.

**Comment by Cavekeeper** [comment]
@Tho85 Could you please more explain

> I think the best solution if you want to trigger something in HA from a F@H binary sensor is to expose the HA entity via emulated_hue, e.g. as a light bulb.

**Comment by Tho85** [comment]
Take a look at the [emulated_hue](https://www.home-assistant.io/integrations/emulated_hue/) component for Home Assistant. If you install it, Home Assistant simulates a Philips Hue bridge that exposes selected lights in Home Assistant as Philips Hue light bulbs.

The free@home SysAP should discover that bridge and connect to it. You should then see a list of all exposed lights under your devices list. You can then drop these virtual Philips Hue bulbs onto your floor plan and connect them like any other free@home light bulb.

IIRC this does not only work for lights in Home Assistant, but for scenes, scripts, automations, etc. as well. So you can trigger almost any entity in Home Assistant through your free@home switches, motion detectors etc.

**Comment by SuperMaximus1984** [comment]
Dears, Is there any update on reading Binary Sensor status from 8/8 Sensor/Actuator unit?
https://www.busch-jaeger.de/en/online-catalogue/product-overview?tx_nlbjproducts_catalog%5Baction%5D=show&tx_nlbjproducts_catalog%5BcatBjeProdukt%5D=15137&tx_nlbjproducts_catalog%5Bcontroller%5D=CatStdArtikel&cHash=b958c2a9b1a888a552d7532a4bd59e0a

As it declares itself as one unit, and in @jheling 's component it can be selected either as Sensor OR Actuator, it can't be recognized by HA as a multi-functional device that contains both. I was able to set & control it as Actuator, but Binary Sensors - unfortunately - no... Any help on how to realize it is appreciated.

**Comment by Tho85** [comment]
@SuperMaximus1984 Support for this device should be improved in #73, you should be able to use the 8 sensor and 8 actuator channels separately. Feel free to try the code there and give feedback.

**Comment by SuperMaximus1984** [comment]
@Tho85 
Thanks! I just don't quite get what files from  @jheling 's original one's I need to replace with yours in order to use reworked code.
Could you please point out?

**Comment by jheling** [comment]
I have just merged @Tho85 pull request. You can just download whole zip and install it on your system. Almost all the files have been touched.
